src/resolvers/archDockerfile.ts

```diff
@@ -13,23 +29,26 @@ type ArchSpecificDockerfile = [string, FileInfo];
 export class ArchDockerfileResolver implements Resolver {
 	public priority = 3;
 	public name = 'Architecture-specific Dockerfile';
-	public allowSpecifiedDockerfile = true;
 	public dockerfileContents: string;
 
 	private archDockerfiles: ArchSpecificDockerfile[] = [];
 
 	public entry(file: FileInfo): void {
 		// We know that this file is a Dockerfile, so just get the extension,
 		// and save it for resolving later
-		const ext = path.extname(file.name).substr(1);
+		const ext = posix.extname(file.name).substr(1);
 		this.archDockerfiles.push([ext, file]);
 	}
 
-	public needsEntry(filepath: string): boolean {
-		const filename = path.basename(filepath);
-		return (
-			filename.startsWith('Dockerfile.') && !filename.endsWith('.template')
-		);
+	public needsEntry(
+		entryPath: ParsedPathPlus,
+		specifiedDockerfilePath?: string,
+	): boolean {
+		const nameMatches = specifiedDockerfilePath
+			? entryPath.unparsed === specifiedDockerfilePath
+			: entryPath.minusExt === 'Dockerfile';
+
+		return nameMatches && !!entryPath.ext && entryPath.ext !== '.template';
```

Comment:
Two cases:
* If a specifiedDockeriflePath was specified, say `'service/MyDockerfile.armv7hf'`, then this resolver (`ArchDockerfileResolver`) will only match a tar entry whose full path is identical to that (`entryPath.unparsed === specifiedDockerfilePath`), and provided that the `specifiedDockeriflePath` has a non-empty file extension different to `'.template'`.
* Otherwise, it will match `Dockerfile.xxx` where xxx is a non-empty file extension different to `'.template'` (as per previous logic). This will only be matched at the root of the project directory tree, as `entryPath.minusExt` is the full path minus the file extension, which must be exactly `'Dockerfile'`. This root-only rule was previously enforced in `resolveTarStreamOnEntry()` through a test: `if (path.normalize(name).split(path.sep).length === 1)`. That test had been added by a recent-ish pull request:
_fix: Correctly resolve contexts with multiple bundle descriptions_
https://github.com/balena-io-modules/resin-bundle-resolve/pull/32/files

Note that both `entryPath` and `specifiedDockerfilePath` are normalized through `TarUtils.normalizeTarEntry()` before the call to `needsEntry()`, so they won't have leading or trailing slashes or redundant path components.